A module for managing ZFS admin privileges

natefoo opened the pull request:
##### SUMMARY
This module manages [ZFS's administrative privileges](http://docs.oracle.com/cd/E18752_01/html/819-5461/gbchv.html).

##### ISSUE TYPE
 - New Module Pull Request

##### COMPONENT NAME
zfs_delegate_admin

##### ANSIBLE VERSION
```
ansible 2.5.0 (zfs-admin-module 4a6119ce56) last updated 2017/10/24 16:25:41 (GMT -400)
  config file = None
  configured module search path = [u'/home/nate/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /home/nate/ansible/ansible/lib/ansible
  executable location = /home/nate/ansible/ansible/bin/ansible
  python version = 2.7.13 (default, Jan 19 2017, 14:48:08) [GCC 6.3.0 20170118]
```

##### ADDITIONAL INFORMATION
Implements support for the `zfs allow` and `zfs unallow` commands, which delegate ZFS administrative privileges (such as the ability to `send`, `recv`, or `create`) on specific ZFS datasets or filesystems to unprivileged users.

natefoo committed:
Create a module for managing ZFS admin privileges

natefoo committed:
Actually support check mode in zfs_permissions

natefoo committed:
Fix Python 2.4 invalid syntax

natefoo committed:
Update added_version to 2.1

natefoo committed:
Remove superflous sys.exit() and add an empty RETURN doc (empty because the module does not return anything other than status-related stuff)

natefoo committed:
Update version_added since zfs_permissions was not accepted in 2.1.

natefoo commented:
Migrated from https://github.com/ansible/ansible-modules-extras/pull/791

xen0l commented:
@natefoo you might also be interested in https://github.com/ansible/ansible/pull/19181.

xen0l commented on lib/ansible/modules/system/zfs_permissions.py:
```diff
@@ -0,0 +1,271 @@
+short_description: Manage zfs administrative permissions
+description:
+  - Manages ZFS file system administrative permissions on Solaris and FreeBSD. See zfs(1M) for more information about the properties.
+version_added: "2.2"
```
This should be 2.3

xen0l commented on lib/ansible/modules/system/zfs_permissions.py:
```diff
@@ -0,0 +1,271 @@
+DOCUMENTATION = '''
+---
+module: zfs_permissions
+short_description: Manage zfs administrative permissions
```
zfs -> ZFS

xen0l commented on lib/ansible/modules/system/zfs_permissions.py:
```diff
@@ -0,0 +1,271 @@
+module: zfs_permissions
+short_description: Manage zfs administrative permissions
+description:
+  - Manages ZFS file system administrative permissions on Solaris and FreeBSD. See zfs(1M) for more information about the properties.
```
Does this also work on Linux? If so, it should be added to the list of supported operating systems. It would be good, if you mentioned OpenZFS support as well.

xen0l commented on lib/ansible/modules/system/zfs_permissions.py:
```diff
@@ -0,0 +1,271 @@
+  name:
+    description:
+      - File system or volume name e.g. C(rpool/myfs)
+    required: true
```
Please, provide type as it might be helpful. I suppose this is type: str.

xen0l commented on lib/ansible/modules/system/zfs_permissions.py:
```diff
@@ -0,0 +1,271 @@
+  users:
+    description:
+      - Users to whom permission(s) should be granted, separated by commas.
+    required: false
```
Please, provide type as it might be helpful. I suppose this is type: str.

dagwieers replied:
Please don't add `required: false` because that is implicit and unnecessary.

senorsmile replied:
I disagree.  Many other modules already shipping have required: false all over the place, and I find this very helpful.  More documentation is never a bad thing (imho).

dagwieers replied:
True, older modules do no comply to the new module contribution documentation because it did not exist. However it is now part of this: http://docs.ansible.com/ansible/dev_guide/developing_modules_checklist.html

BTW It is easier to find which parameters are required, when only the ones that are required have `required: true`. If they all have a `required:` option, you have to find which ones have true on the right line. Not very convenient, hence we have this best practice.

senorsmile replied:
Ah, I hadn't seen those specific recommendations.  Thanks for the link!

xen0l commented on lib/ansible/modules/system/zfs_permissions.py:
```diff
@@ -0,0 +1,271 @@
+# This module does not return anything other than the standard
+# changed/state/msg/stdout
+RETURN = '''
+
```
Remove empty line.

xen0l commented on lib/ansible/modules/system/zfs_permissions.py:
```diff
@@ -0,0 +1,271 @@
+        self.module = module
+        self.name = module.params.get('name')
+        self.state = module.params.get('state')
+        self.users = module.params.get('users') or []
```
get supports default type (https://docs.python.org/2/library/stdtypes.html#dict.get), so you can use _module.prams.get('users', [])_ instead of _module.params.get('users') or []_

xen0l commented on lib/ansible/modules/system/zfs_permissions.py:
```diff
@@ -0,0 +1,271 @@
+
+    def run_command(self, cmd):
+        progname = cmd[0]
+        cmd[0] = self.module.get_bin_path(progname, True)
```
You are looking for zfs(1M) everytime you run this function. Why not do it in __init__ when initiating class and using a variable, e.g. self._zfs_binary?

xen0l commented on lib/ansible/modules/system/zfs_permissions.py:
```diff
@@ -0,0 +1,271 @@
+      - Apply permissions to everyone.
+    required: false
+    default: false
+    choices: ['on','off']
```
You are saying that on/off is the accepted values, yet in argument_spec you are using BOOLEANS, which you are converting with module.boolean. Why not accepting booleans here and remove the call to module.boolean()? Also defining type: bool for everyone/local/descendents/recursive would be more consistent with other modules.

xen0l commented on lib/ansible/modules/system/zfs_permissions.py:
```diff
@@ -0,0 +1,271 @@
+            for perm in self.perms:
+                if ((self.state == 'present' and perm not in everyone_perms) or
+                    (self.state == 'absent' and perm in everyone_perms)):
+                    #
```
Missing comment?

xen0l commented:
@natefoo you might also want to move this module to storage/zfs, so all ZFS related modules are displayed in its own category in the documentation, which helps a lot when referring to documentation.

tmshn commented on lib/ansible/modules/system/zfs_permissions.py:
```diff
@@ -0,0 +1,271 @@
+            local = dict(default=None, choices=BOOLEANS),
+            descendents = dict(default=None, choices=BOOLEANS),
+            recursive = dict(default=False, choices=BOOLEANS)
+        ),
```
You can use `type='bool'` rather than `choices=BOOLEANS`, with which the module automatically do type cast `module.boolean(module.params.get(...))` for you.
https://github.com/natefoo/ansible/blob/599eb6115e53fd6fd15770a4c7ef292177d5b888/lib/ansible/module_utils/basic.py#L1581

(Note that `None` will be passed as is; without conversion. See https://github.com/natefoo/ansible/blob/599eb6115e53fd6fd15770a4c7ef292177d5b888/lib/ansible/module_utils/basic.py#L1654-L1655)

Similary, you can use `type='list' default=list()` for `users`, `groups` and `permissions`.
https://github.com/natefoo/ansible/blob/599eb6115e53fd6fd15770a4c7ef292177d5b888/lib/ansible/module_utils/basic.py#L1525

tmshn commented:
I don't know if it's required, but some modules have `ANSIBLE_METADATA`.
How about adding this?

E.g.;
https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/system/timezone.py#L31-L33

dagwieers commented on lib/ansible/modules/system/zfs_permissions.py:
```diff
@@ -0,0 +1,271 @@
+  groups:
+    description:
+      - Groups to whom permission(s) should be granted, separated by commas.
+    required: false
```
Same. Please remove.

dagwieers commented on lib/ansible/modules/system/zfs_permissions.py:
```diff
@@ -0,0 +1,271 @@
+  everyone:
+    description:
+      - Apply permissions to everyone.
+    required: false
```
Same. Please remove.

dagwieers commented on lib/ansible/modules/system/zfs_permissions.py:
```diff
@@ -0,0 +1,271 @@
+  permissions:
+    description:
+      - The permission(s) to delegate, separated by commas (required if C(state) is C(present))
+    required: false
```
Same. Please remove.

dagwieers commented on lib/ansible/modules/system/zfs_permissions.py:
```diff
@@ -0,0 +1,271 @@
+  local:
+    description:
+      - Apply permissions to C(name) "locally" (C(zfs allow -l))
+    required: false
```
Same. Please remove.

dagwieers commented on lib/ansible/modules/system/zfs_permissions.py:
```diff
@@ -0,0 +1,271 @@
+  descendents:
+    description:
+      - Apply permissions to C(name)'s descendents (C(zfs allow -d))
+    required: false
```
Same. Please remove.

dagwieers commented on lib/ansible/modules/system/zfs_permissions.py:
```diff
@@ -0,0 +1,271 @@
+  recursive:
+    description:
+      - Unallow permissions recursively (ignored when C(state) is C(present))
+    required: false
```
Same. Please remove.

dagwieers commented on lib/ansible/modules/system/zfs_permissions.py:
```diff
@@ -0,0 +1,271 @@
+
+
+# import module snippets
+from ansible.module_utils.basic import *
```
Please move these to the top of the module and only import the stuff you are using from the library.
```python
from ansible.module_utils.basic import AnsibleModule
```

natefoo commented:
Thanks for the reviews, all. This module was (iirc) based off the zfs module and carried over some out of date practices. I'll incorporate the feedback ASAP.

dagwieers commented:
Please read: http://docs.ansible.com/ansible/dev_guide/developing_modules_checklist.html

Akasurde commented on lib/ansible/modules/system/zfs_permissions.py:
```diff
@@ -0,0 +1,271 @@
+        self.changed = False
+        self.__current_perms = None
+
+        if self.state == 'present' and not self.perms:
```
use `required_if`

Akasurde commented on lib/ansible/modules/system/zfs_permissions.py:
```diff
@@ -0,0 +1,271 @@
+        if self.state == 'present' and not self.perms:
+            self.module.fail_json(msg='The `permissions` option is required for state=present')
+
+        if self.state == 'present' and not (self.users or self.groups or self.everyone):
```
use `required_if`

natefoo replied:
This one wasn't possible because there's not a `required_if_any_of`.

natefoo committed:
- Rename zfs_permissions to zfs_delegate_admin to more accurately reflect
  its purpose.
- Update with current Ansible module best practices.
- Stop checking whether changes need to be made (this was flawed since
  it was not able to detect whether, for example, changes would be made
  with `local=True` when the permissions already existed for
  `Local+Descendent`. Instead, just make the changes (`zfs allow` is
  idempotent) and compare the state before and after.
- Generally write things in a better way.

natefoo committed:
Fix test errors

natefoo committed:
Fix more tests.

natefoo commented:
ready_for_review

natefoo commented:
+affects_2.5
-affects_2.3

natefoo commented:
Can someone with the privileges fix the labels on this, since @ansibot isn't able to do so?

natefoo committed:
Catch exceptions when parsing `zfs allow` output. At least on one
version of ZFS (on a 2016 version of illumos) will have a single space
in the user or group column if the name can't be resolved by the
system's name services).

maxamillion commented:
@dagwieers @Akasurde re-ping

dagwieers commented:
rebuild_merge

dagwieers commented:
This one has been in the queue for too long, and only minor changes are needed.
I am going to merge this and fix them in a subsequent commit.

mattclay commented:
@dagwieers Please don't merge PRs that have not passed tests or are failing tests. If fixes need to be made, they need to be done as part of the PR before merging. If that's not possible, close the original PR and open a new one with the original work plus the necessary fixes.

gundalow commented:
Examples need updating to include, `- name:` and use `key: value`

dagwieers commented:
@gundalow This was done in the subsequent PR. The author did not allow me to add commits to this PR.
